kubeadm: add task pages for adding Linux and Windows worker nodes (#47888)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubeadm/adding-linux-nodes.md

---
title: Adding Linux worker nodes
content_type: task
weight: 50
---

<!-- overview -->

This page explains how to add Linux worker nodes to a kubeadm cluster.

## {{% heading "prerequisites" %}}

* Each joining worker node has installed the required components from
[Installing kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/), such as,

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Reviewer: This is the part to repeat (I've edited it, eg, to drop the 2 CPU requirement, which isn't true for worker nodes)

> - 2 GiB or more of RAM per server; any less will leave little room for your apps.
> - Full network connectivity between all machines in the cluster (public or private network is fine).
> - Unique hostname, MAC address, and product_uuid (unique for every node). See [here](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#verify-mac-address) for more details.
> - All the [required network ports](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#check-required-ports) allow connections from the control plane.

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Author: but the user should navigate to the install kubeadm page if they haven't installed all the requirement components.

> (I've edited it, eg, to drop the 2 CPU requirement, which isn't true for worker nodes)

this can be clarified on the install page:
>> 2 CPUs or more for control plane machines.

> BTW I'm not sure if kubeadm actually requires that nodes have Ethernet.

the user can choose to deploy or local network if they like

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Reviewer: > the install kubeadm page

That's giving the prerequisites for a control plane, right? We could use tabs there to clarify the different requirements.
But yes, we can can explain that you need to have already installed `kubeadm`.

IMO a later PR should refactor it so we _don't_ send people there; the way you install `kubeadm` is not that similar between Windows and Linux.

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Author: added:
> 2 CPUs or more for control plane machines.

to the install page in the latest commit.

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Author: > That's giving the prerequisites for a control plane, right? 

no, that section was for machines in general CP or workers, so the `2CPU for CP` was needed.

> kubeadm, the kubelet and a {{< glossary_tooltip term_id="container-runtime" text="container runtime" >}}.

Reviewer: OK, this'll do!

1. Become root (e.g. `sudo su -`).
1. Run the command that was output by `kubeadm init`. For example:

  ```bash
  kubeadm join --token <token> <control-plane-host>:<control-plane-port> --discovery-token-ca-cert-hash sha256:<hash>
  ```

### Additional information for kubeadm join

{{< note >}}
To specify an IPv6 tuple for `<control-plane-host>:<control-plane-port>`, IPv6 address must be enclosed in square brackets, for example: `[2001:db8::101]:2073`.
{{< /note >}}

If you do not have the token, you can get it by running the following command on the control plane node:

```bash
kubeadm token list
```

The output is similar to this:

```console
TOKEN                    TTL  EXPIRES              USAGES           DESCRIPTION            EXTRA GROUPS
8ewj1p.9r9hcjoqgajrj4gi  23h  2018-06-12T02:51:28Z authentication,  The default bootstrap  system:
                                                   signing          token generated by     bootstrappers:
                                                                    'kubeadm init'.        kubeadm:
                                                                                           default-node-token
```

> By default, tokens expire after 24 hours. If you are joining a node to the cluster after the current token has expired. You can create a new token by running the following command on the control plane node:

Reviewer: ```suggestion
By default, node join tokens expire after 24 hours. If you are joining a node to the cluster after the
current token has expired, you can create a new token by running the following command on the
control plane node:
```

```bash

> kubeadm token create

Reviewer: ```suggestion
# Run this on a control plane node
sudo kubeadm token create
```

Author: updated, minus `# Run this on a control plane node`. the prev sentence already mentions `...running the following command on the control plane node`

```

The output is similar to this:

```console
5didvk.d09sbcov8ph2amjw
```

If you don't have the value of `--discovery-token-ca-cert-hash`, you can get it by running the
following commands on the control plane node:

```bash

> openssl x509 -pubkey -in /etc/kubernetes/pki/ca.crt | openssl rsa -pubin -outform der 2>/dev/null | \

Reviewer: ```suggestion
# Run this on a control plane node
sudo cat /etc/kubernetes/pki/ca.crt | openssl x509 -pubkey  | openssl rsa -pubin -outform der 2>/dev/null | \
```
(least privilege)

Author: updated, minus the `# Run this on a control plane node` part that is already in the prev sentence.

Reviewer: It doesn't block a merge, but experience confidently shows us that people don't read documentation and many people paste the commands in without reading the text in between.

Author: well, it will fail on a joining worker since the ca.crt doens't exist yet.
so then they will go back and read the prev sentence.

   openssl dgst -sha256 -hex | sed 's/^.* //'
```

The output is similar to:

```console
8cb2de97839780a412b93877f8507ad6c94f73add17d5d7058e91741c9d5ec78
```

The output of the `kubeadm join` command should look something like:

```
[preflight] Running pre-flight checks

... (log output of join workflow) ...

Node join complete:
* Certificate signing request sent to control-plane and response
  received.
* Kubelet informed of new secure connection details.

Run 'kubectl get nodes' on control-plane to see this machine join.
```

A few seconds later, you should notice this node in the output from `kubectl get nodes`
when run on the control plane node.

{{< note >}}
As the cluster nodes are usually initialized sequentially, the CoreDNS Pods are likely to all run
on the first control plane node. To provide higher availability, please rebalance the CoreDNS Pods
with `kubectl -n kube-system rollout restart deployment coredns` after at least one new node is joined.
{{< /note >}}

## {{% heading "whatsnext" %}}

> * See how to [Adding Windows worker nodes](/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).

Reviewer: ```suggestion
* See how to [add Windows worker nodes](/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).
```
